// @colony/colony-js-adapter-ethers/lib/EthersAdapter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _ethers = require('ethers');

var _ethers2 = _interopRequireDefault(_ethers);

var _colonyJsUtils = require('@colony/colony-js-utils');

var _colonyJsContractLoader = require('@colony/colony-js-contract-loader');

var _colonyJsContractLoader2 = _interopRequireDefault(_colonyJsContractLoader);

var _EthersContract = require('./EthersContract');

var _EthersContract2 = _interopRequireDefault(_EthersContract);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var EthersAdapter = function () {
  (0, _createClass3.default)(EthersAdapter, null, [{
    key: 'getEventPromises',
    value: function getEventPromises(_ref) {
      var events = _ref.events,
          timeoutMs = _ref.timeoutMs,
          transactionHash = _ref.transactionHash;

      var mapEventToPromise = function mapEventToPromise(eventName) {
        var contract = void 0;
        var handler = void 0;
        return (0, _colonyJsUtils.raceAgainstTimeout)(new Promise(function (resolve) {
          var _events$eventName = events[eventName];
          contract = _events$eventName.contract;
          handler = _events$eventName.handler;

          contract.addListener(eventName, transactionHash, function (_ref2) {
            var args = _ref2.args;
            return resolve(handler(args));
          });
        }), timeoutMs, function () {
          return contract.removeListener(eventName, transactionHash);
        });
      };
      return Object.getOwnPropertyNames(events).map(mapEventToPromise);
    }
  }]);

  function EthersAdapter(_ref3) {
    var loader = _ref3.loader,
        provider = _ref3.provider,
        wallet = _ref3.wallet;
    (0, _classCallCheck3.default)(this, EthersAdapter);

    this.loader = loader;
    this.provider = provider;
    this.wallet = wallet;
  }

  (0, _createClass3.default)(EthersAdapter, [{
    key: 'getContract',
    value: function () {
      var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(query) {
        var _ref5, address, abi;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.loader.load(query, {
                  abi: true,
                  address: true,
                  bytecode: false
                });

              case 2:
                _ref5 = _context.sent;
                address = _ref5.address;
                abi = _ref5.abi;

                if (!(typeof address !== 'string')) {
                  _context.next = 7;
                  break;
                }

                throw new Error('Unable to parse contract address');

              case 7:
                return _context.abrupt('return', new _EthersContract2.default(address, abi, this.wallet));

              case 8:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getContract(_x) {
        return _ref4.apply(this, arguments);
      }

      return getContract;
    }()
  }, {
    key: 'getContractDeployTransaction',
    value: function () {
      var _ref6 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(query, contractParams) {
        var _ethers$Contract;

        var _ref7, abi, bytecode;

        return _regenerator2.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.next = 2;
                return this.loader.load(query, {
                  abi: true,
                  address: false,
                  bytecode: true
                });

              case 2:
                _ref7 = _context2.sent;
                abi = _ref7.abi;
                bytecode = _ref7.bytecode;
                return _context2.abrupt('return', (_ethers$Contract = _ethers2.default.Contract).getDeployTransaction.apply(_ethers$Contract, [bytecode, abi].concat((0, _toConsumableArray3.default)(contractParams))));

              case 6:
              case 'end':
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getContractDeployTransaction(_x2, _x3) {
        return _ref6.apply(this, arguments);
      }

      return getContractDeployTransaction;
    }()
    // XXX this isn't a static method because we can't define it as such
    // in the Interface thanks to Flow

  }, {
    key: 'getEventData',
    value: function () {
      var _ref9 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(_ref8) {
        var _ref8$events = _ref8.events,
            events = _ref8$events === undefined ? {} : _ref8$events,
            transactionHash = _ref8.transactionHash,
            timeoutMs = _ref8.timeoutMs;
        var eventPromises;
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                eventPromises = this.constructor.getEventPromises({
                  events: events,
                  timeoutMs: timeoutMs,
                  transactionHash: transactionHash
                });
                _context3.prev = 1;
                _context3.t0 = Object.assign;
                _context3.t1 = Object;
                _context3.t2 = [{}];
                _context3.t3 = _toConsumableArray3.default;
                _context3.next = 8;
                return Promise.all(eventPromises);

              case 8:
                _context3.t4 = _context3.sent;
                _context3.t5 = (0, _context3.t3)(_context3.t4);
                _context3.t6 = _context3.t2.concat.call(_context3.t2, _context3.t5);
                return _context3.abrupt('return', _context3.t0.apply.call(_context3.t0, _context3.t1, _context3.t6));

              case 12:
                _context3.prev = 12;

                Object.entries(events).forEach(function (_ref10) {
                  var _ref11 = (0, _slicedToArray3.default)(_ref10, 2),
                      eventName = _ref11[0],
                      contract = _ref11[1].contract;

                  contract.removeListener(eventName, transactionHash);
                });
                return _context3.finish(12);

              case 15:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this, [[1,, 12, 15]]);
      }));

      function getEventData(_x4) {
        return _ref9.apply(this, arguments);
      }

      return getEventData;
    }()
  }, {
    key: 'getTransactionReceipt',
    value: function () {
      var _ref12 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(transactionHash) {
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                return _context4.abrupt('return', this.provider.getTransactionReceipt(transactionHash));

              case 1:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getTransactionReceipt(_x5) {
        return _ref12.apply(this, arguments);
      }

      return getTransactionReceipt;
    }()
    /**
     * Sign a message hash (as binary) and return a split signature.
     */

  }, {
    key: 'signMessage',
    value: function () {
      var _ref13 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5(messageHash) {
        var messageBytes, signature, _ethers$utils$splitSi, sigR, sigS, sigV;

        return _regenerator2.default.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                messageBytes = _ethers2.default.utils.arrayify(messageHash);
                _context5.next = 3;
                return this.wallet.signMessage(messageBytes);

              case 3:
                signature = _context5.sent;
                _ethers$utils$splitSi = _ethers2.default.utils.splitSignature(signature), sigR = _ethers$utils$splitSi.r, sigS = _ethers$utils$splitSi.s, sigV = _ethers$utils$splitSi.v;
                return _context5.abrupt('return', {
                  sigR: sigR,
                  sigS: sigS,
                  sigV: sigV
                });

              case 6:
              case 'end':
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function signMessage(_x6) {
        return _ref13.apply(this, arguments);
      }

      return signMessage;
    }()

    /**
     * Given a message digest and a signature, recover the address used to
     * sign the message.
     */
    // eslint-disable-next-line class-methods-use-this

  }, {
    key: 'ecRecover',
    value: function ecRecover(digest, _ref14) {
      var sigR = _ref14.sigR,
          sigS = _ref14.sigS,
          sigV = _ref14.sigV;

      // This method doesn't need to be static, but flow Interfaces don't
      // support static methods.
      var recoveryParam = sigV - 27;
      return _ethers2.default.SigningKey.recover(digest, sigR, sigS, recoveryParam);
    }
  }]);
  return EthersAdapter;
}();

exports.default = EthersAdapter;
